llvm/lib/Transforms/Instrumentation/InstrProfiling.cpp

```diff
@@ -1213,8 +1213,20 @@ Value *InstrLowerer::getBitmapAddress(InstrProfMCDCTVBitmapUpdate *I) {
 void InstrLowerer::lowerCover(InstrProfCoverInst *CoverInstruction) {
   auto *Addr = getCounterAddress(CoverInstruction);
   IRBuilder<> Builder(CoverInstruction);
-  // We store zero to represent that this block is covered.
-  Builder.CreateStore(Builder.getInt8(0), Addr);
+  if (ConditionalCounterUpdate) {
+    auto &Ctx = CoverInstruction->getParent()->getContext();
+    auto *Int8Ty = llvm::Type::getInt8Ty(Ctx);
+    Value *Load = Builder.CreateLoad(Int8Ty, Addr, "pgocount");
+    Value *Cmp = Builder.CreateICmpNE(Load, ConstantInt::get(Int8Ty, 0),
+                                      "pgocount.ifnonzero");
+    Value *Sel =
+        Builder.CreateSelect(Cmp, Builder.getInt8(0), Load, "pgocount.select");
```

Comment:
I think this has to be a compare and branch instruction instead of a load.

When I tried using a select statement like what's being done here, the select ends up being optimized to an unconditional write, which defeats this feature.

Source: see my earlier post on Discourse: https://discourse.llvm.org/t/rfc-single-byte-counters-for-source-based-code-coverage/75685/12?u=ayzhao

Reply:
Ok, that's interesting. Thanks for pointing that out, and I replaced that with a compare and branch.
https://discourse.llvm.org/t/rfc-single-byte-counters-for-source-based-code-coverage/75685/10?u=gulfemsavrun

I realized that you added a break with the following explaination:
```
       } else if (auto *IPC = dyn_cast<InstrProfCoverInst>(&Instr)) {
         lowerCover(IPC);
         MadeChange = true;
+        break; // because this splits and terminates the current basic block
```

Is this really necessary? I did not find it necessary for the simple testing that I did, but am I missing something?